usage/bigquery.go

```diff
@@ -0,0 +1,161 @@
+	UserID     string
+}
+
+type QuerySpec struct {
+	From, To *time.Time
+	Filter   QueryFilter
+}
+
+type UsageSummaryRow struct {
+	UserID    string `bigquery:"user_id"`
+	CreatorID string `bigquery:"creator_id"`
+
+	ViewCount       int64              `bigquery:"view_count"`
+	LegacyViewCount bigquery.NullInt64 `bigquery:"legacy_view_count"`
+	PlaytimeMins    float64            `bigquery:"playtime_mins"`
+}
+
+type BigQuery interface {
+	QueryUsageSummary(ctx context.Context, userID string, creatorID string, spec QuerySpec) (*UsageSummaryRow, error)
+}
+
+type BigQueryOptions struct {
+	BigQueryCredentialsJSON   string
+	HourlyUsageTable          string
+	MaxBytesBilledPerBigQuery int64
+}
+
+func NewBigQuery(opts BigQueryOptions) (BigQuery, error) {
+	bigquery, err := bigquery.NewClient(context.Background(),
+		bigquery.DetectProjectID,
+		option.WithCredentialsJSON([]byte(opts.BigQueryCredentialsJSON)))
+	if err != nil {
+		return nil, fmt.Errorf("error creating bigquery client: %w", err)
+	}
+
+	return &bigqueryHandler{opts, bigquery}, nil
+}
+
+// interface from *bigquery.Client to allow mocking
+type bigqueryClient interface {
+	Query(q string) *bigquery.Query
+}
+
+type bigqueryHandler struct {
+	opts   BigQueryOptions
+	client bigqueryClient
+}
+
+// usage summary query
+
+func (bq *bigqueryHandler) QueryUsageSummary(ctx context.Context, userID string, creatorID string, spec QuerySpec) (*UsageSummaryRow, error) {
+	sql, args, err := buildUsageSummaryQuery(bq.opts.HourlyUsageTable, userID, creatorID, spec)
+	if err != nil {
+		return nil, fmt.Errorf("error building usage summary query: %w", err)
+	}
+
+	bqRows, err := doBigQuery[UsageSummaryRow](bq, ctx, sql, args)
+	if err != nil {
+		return nil, fmt.Errorf("bigquery error: %w", err)
+	} else if len(bqRows) > 1 {
+		return nil, fmt.Errorf("internal error, query returned %d rows", len(bqRows))
+	}
+
+	if len(bqRows) == 0 {
+		return nil, nil
+	}
+	return &bqRows[0], nil
+}
+
+func buildUsageSummaryQuery(table string, userID string, creatorID string, spec QuerySpec) (string, []interface{}, error) {
+	if userID == "" {
+		return "", nil, fmt.Errorf("userID cannot be empty")
+	}
+
+	query := squirrel.Select(
+		"cast(sum(transcode_total_usage_minutes) as FLOAT64) as transcode_total_usage_minutes",
+		"cast(sum(delivery_usage_gbs) as FLOAT64) as delivery_usage_gbs",
+		"cast(avg(storage_usage_gbs) as FLOAT64) as storage_usage_gbs").
+		From(table).
+		Limit(2)
+
+	if creatorId := spec.Filter.CreatorID; creatorId != "" {
+		query = query.Where("creator_id_type = ?", "unverified")
+		query = query.Where("creator_id = ?", creatorID)
+	}
+
+	query = withUserIdFilter(query, userID)
+
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return "", nil, err
+	}
+
+	return sql, args, nil
+}
+
+// query helpers
+
+func withUserIdFilter(query squirrel.SelectBuilder, userID string) squirrel.SelectBuilder {
+	if userID == "" {
+		query = query.Column("user_id").GroupBy("user_id")
+	} else {
```

Comment:
Do we have any checks against who could call this API without a `userId`? I don't think any user should be able to access other users data.

Reply:
The userID used in the query is coming from authentication, which actually now that you mention it, it's not ok. This is supposed to be a call that we make to calculate billing, so this is probably gonna be called by Studio and not directly by users

Reply:
If it will always be there, let's disallow the query from executing if no user ID is provided. Just simplify this logic and make it always do the `Where`. If no user ID is provided it will filter for an empty user ID, which is better than the risk of returning cross-user data on this endpoint if something goes wrong in the future.